### ISSUE TITLE
Cypher query not following `scriptEvaluationTimeout`

### ISSUE THREAD

**Issue by SarthakGhosh16** [opencypher/cypher-for-gremlin#335]
### Description
Issue regarding: 
1. https://github.com/opencypher/cypher-for-gremlin/pull/316
2. https://github.com/opencypher/cypher-for-gremlin/pull/315

I updated my pom.xml with 1.0.3 version 
```
<dependency>
	<groupId>org.opencypher.gremlin</groupId>
	<artifactId>cypher-gremlin-server-client</artifactId>
	<version>1.0.3</version>
</dependency>
```

These are the jars I have:
cypher-gremlin-server-client-1.0.3.jar
cypher-gremlin-extensions-1.0.3.jar

When running a very generic query like (match p=(a)-[]-(b) return p), the gremlin server goes into a hung state. Shouldn't client drop the connection after `scriptEvaluationTimeout` has been exceeded? 

I'm connecting via **`remote-objects.yaml`** file
```
hosts: [host]
port: 8182
serializer: { className: org.apache.tinkerpop.gremlin.driver.ser.GryoMessageSerializerV3d0, config: { ioRegistries: [org.janusgraph.graphdb.tinkerpop.JanusGraphIoRegistry,org.apache.tinkerpop.gremlin.tinkergraph.structure.TinkerIoRegistryV3d0] }}
connectionPool: {
  enableSsl: false,
  maxInProcessPerConnection: 16,
  maxSimultaneousUsagePerConnection: 32,
  maxSize: 32,
  maxContentLength: 81928192
}
workerPoolSize: 16
nioPoolSize: 8
```

Jar at **gremlin server v3.3.9:** 
cypher-gremlin-server-plugin-0.9.13-all.jar
I've configured through Manual Installation (https://github.com/opencypher/cypher-for-gremlin/tree/master/tinkerpop/cypher-gremlin-server-plugin#manual-installation)

**gremlin-server.yaml**
scriptEvaluationTimeout: 120000
staticImports: ['org.opencypher.gremlin.process.traversal.CustomPredicates.*']}}
processors:
  - { className: org.apache.tinkerpop.gremlin.server.op.session.SessionOpProcessor, config: { sessionTimeout: 28800000 }}
  - { className: org.apache.tinkerpop.gremlin.server.op.traversal.TraversalOpProcessor, config: { cacheExpirationTime: 600000, cacheMaxSize: 1000 }}
  - { className: org.opencypher.gremlin.server.op.cypher.CypherOpProcessor, config: { sessionTimeout: 28800000}}

**Comment by SarthakGhosh16** [comment]
@dwitry 
I also tried a query where gremlin was still available and returned data but it took more than 10 minutes.
Log:
```
eDecoder - Decoding WebSocket Frame opCode=2
[11/11/19 19:00:55:930 UTC] 000007e1 SystemOut                                                    O 19:00:55.929 [gremlin-driver-loop-1] DEBUG i.n.h.c.h.w.WebSocket08FrameDecoder - Decoding WebSocket Frame length=23846
[11/11/19 19:01:36:238 UTC] 000007e1 SystemOut                                                    O 19:01:36.238 [gremlin-driver-loop-1] DEBUG i.n.h.c.h.w.WebSocket08FrameDecoder - Decoding WebSocket Frame opCode=2
[11/11/19 19:01:36:240 UTC] 000007e1 SystemOut                                                    O 19:01:36.240 [gremlin-driver-loop-1] DEBUG i.n.h.c.h.w.WebSocket08FrameDecoder - Decoding WebSocket Frame length=23849
[11/11/19 19:02:35:369 UTC] 000007e1 SystemOut                                                    O 19:02:35.369 [gremlin-driver-loop-1] DEBUG i.n.h.c.h.w.WebSocket08FrameDecoder - Decoding WebSocket Frame opCode=2
[11/11/19 19:02:35:371 UTC] 000007e1 SystemOut                                                    O 19:02:35.370 [gremlin-driver-loop-1] DEBUG i.n.h.c.h.w.WebSocket08FrameDecoder - Decoding WebSocket Frame length=23852
[11/11/19 19:03:05:104 UTC] 000007e1 SystemOut                                                    O 19:03:05.103 [gremlin-driver-loop-1] DEBUG i.n.h.c.h.w.WebSocket08FrameDecoder - Decoding WebSocket Frame opCode=2
[11/11/19 19:03:05:105 UTC] 000007e1 SystemOut                                                    O 19:03:05.105 [gremlin-driver-loop-1] DEBUG i.n.h.c.h.w.WebSocket08FrameDecoder - Decoding WebSocket Frame length=23914
[11/11/19 19:03:37:692 UTC] 000007e1 SystemOut                                                    O 19:03:37.692 [gremlin-driver-loop-1] DEBUG i.n.h.c.h.w.WebSocket08FrameDecoder - Decoding WebSocket Frame opCode=2
[11/11/19 19:03:37:694 UTC] 000007e1 SystemOut                                                    O 19:03:37.693 [gremlin-driver-loop-1] DEBUG i.n.h.c.h.w.WebSocket08FrameDecoder - Decoding WebSocket Frame length=23920
[11/11/19 19:03:53:827 UTC] 000007e1 SystemOut                                                    O 19:03:53.826 [gremlin-driver-loop-1] DEBUG i.n.h.c.h.w.WebSocket08FrameDecoder - Decoding WebSocket Frame opCode=2
[11/11/19 19:03:53:828 UTC] 000007e1 SystemOut                                                    O 19:03:53.828 [gremlin-driver-loop-1] DEBUG i.n.h.c.h.w.WebSocket08FrameDecoder - Decoding WebSocket Frame length=23922
[11/11/19 19:04:32:504 UTC] 000007e1 SystemOut                                                    O 19:04:32.504 [gremlin-driver-loop-1] DEBUG i.n.h.c.h.w.WebSocket08FrameDecoder - Decoding WebSocket Frame opCode=2
[11/11/19 19:04:32:505 UTC] 000007e1 SystemOut                                                    O 19:04:32.505 [gremlin-driver-loop-1] DEBUG i.n.h.c.h.w.WebSocket08FrameDecoder - Decoding WebSocket Frame length=23928
[11/11/19 19:04:55:742 UTC] 000007e1 SystemOut                                                    O 19:04:55.741 [gremlin-driver-loop-1] DEBUG i.n.h.c.h.w.WebSocket08FrameDecoder - Decoding WebSocket Frame opCode=2
[11/11/19 19:04:55:743 UTC] 000007e1 SystemOut                                                    O 19:04:55.743 [gremlin-driver-loop-1] DEBUG i.n.h.c.h.w.WebSocket08FrameDecoder - Decoding WebSocket Frame length=23906
[11/11/19 19:05:35:660 UTC] 000007e1 SystemOut                                                    O 19:05:35.660 [gremlin-driver-loop-1] DEBUG i.n.h.c.h.w.WebSocket08FrameDecoder - Decoding WebSocket Frame opCode=2
[11/11/19 19:05:35:662 UTC] 000007e1 SystemOut                                                    O 19:05:35.662 [gremlin-driver-loop-1] DEBUG i.n.h.c.h.w.WebSocket08FrameDecoder - Decoding WebSocket Frame length=23854
[11/11/19 19:07:12:869 UTC] 000007e1 SystemOut                                                    O 19:07:12.869 [gremlin-driver-loop-1] DEBUG i.n.h.c.h.w.WebSocket08FrameDecoder - Decoding WebSocket Frame opCode=2
[11/11/19 19:07:12:871 UTC] 000007e1 SystemOut                                                    O 19:07:12.870 [gremlin-driver-loop-1] DEBUG i.n.h.c.h.w.WebSocket08FrameDecoder - Decoding WebSocket Frame length=23834
[11/11/19 19:08:04:332 UTC] 000007e1 SystemOut                                                    O 19:08:04.332 [gremlin-driver-loop-1] DEBUG i.n.h.c.h.w.WebSocket08FrameDecoder - Decoding WebSocket Frame opCode=2
[11/11/19 19:08:04:333 UTC] 000007e1 SystemOut                                                    O 19:08:04.333 [gremlin-driver-loop-1] DEBUG i.n.h.c.h.w.WebSocket08FrameDecoder - Decoding WebSocket Frame length=23791
[11/11/19 19:08:36:298 UTC] 000007e1 SystemOut                                                    O 19:08:36.298 [gremlin-driver-loop-1] DEBUG i.n.h.c.h.w.WebSocket08FrameDecoder - Decoding WebSocket Frame opCode=2
[11/11/19 19:08:36:299 UTC] 000007e1 SystemOut                                                    O 19:08:36.299 [gremlin-driver-loop-1] DEBUG i.n.h.c.h.w.WebSocket08FrameDecoder - Decoding WebSocket Frame length=23897
[11/11/19 19:08:54:000 UTC] 000007e1 SystemOut                                                    O 19:08:54.000 [gremlin-driver-loop-1] DEBUG i.n.h.c.h.w.WebSocket08FrameDecoder - Decoding WebSocket Frame opCode=2
[11/11/19 19:08:54:001 UTC] 000007e1 SystemOut                                                    O 19:08:54.001 [gremlin-driver-loop-1] DEBUG i.n.h.c.h.w.WebSocket08FrameDecoder - Decoding WebSocket Frame length=10469
[11/11/19 19:08:54:007 UTC] 00000a53 SystemOut                                                    O 19:08:54.005 [gremlin-driver-worker-3] DEBUG o.a.t.gremlin.driver.ConnectionPool - Attempting to return Connection{host=Host{address=fci-graph-writer-gremlin/10.96.248.196:8182, hostUri=ws://fci-graph-writer-gremlin:8182/gremlin}} on Host{address=fci-graph-writer-gremlin/10.96.248.196:8182, hostUri=ws://fci-graph-writer-gremlin:8182/gremlin}
[11/11/19 19:08:54:008 UTC] 00000a53 SystemOut                                                    O 19:08:54.008 [gremlin-driver-worker-3] DEBUG o.a.t.gremlin.driver.ConnectionPool - Announce connection available on Host{address=fci-graph-writer-gremlin/10.96.248.196:8182, hostUri=ws://fci-graph-writer-gremlin:8182/gremlin}
[11/11/19 19:08:54:230 UTC] 000007dd SystemOut                                                    O 19:08:54.230 [Default Executor-thread-1849] INFO  o.a.t.gremlin.driver.ConnectionPool - Signalled closing of connection pool on Host{address=fci-graph-writer-gremlin/10.96.248.196:8182, hostUri=ws://fci-graph-writer-gremlin:8182/gremlin} with core size of 2
[11/11/19 19:08:54:232 UTC] 000007e1 SystemOut                                                    O 19:08:54.231 [gremlin-driver-loop-1] DEBUG i.n.h.c.h.w.WebSocket08FrameEncoder - Encoding WebSocket Frame opCode=8 length=0
[11/11/19 19:08:54:232 UTC] 000007e2 SystemOut                                                    O 19:08:54.231 [gremlin-driver-loop-2] DEBUG i.n.h.c.h.w.WebSocket08FrameEncoder - Encoding WebSocket Frame opCode=8 length=0
```

As you can see, the query executed around `[11/11/19 19:00:55:930 UTC]` but retuned data by `[11/11/19 19:08:54:232 UTC]` 
`scriptEvaluationTimeout` is set to 2 minutes.

Do I need to explicitly do something in my Java code?  Other processors cut the connection at 2 minutes and I haven'e called any function explicitly with them either.

**Comment by dwitry** [comment]
Hello @SarthakGhosh16,

according to first comment, you are using cypher-gremlin-server-plugin-**0.9.13**-all.jar. Fix #316 was released in **1.0.3**.

Upgrading Client is not enough for this fix to work. You will need to upgrade Gremlin Server to 3.4.x and plugin to 1.0.3.

**Comment by SarthakGhosh16** [comment]
Hi @dwitry
My gremlin server will be `3.3.x`. All my dependencies at client and server level are compatible with `janusgraph 0.3.1` 
Will there be an update for the versions I have mentioned in the description? 
If not, is there a work around for this?

I have done something like this for now
```

      ExecutorService executor = Executors.newCachedThreadPool();
      Callable<Object> task =
          new Callable<Object>() {
            public Object call() {
              return cF.cypherFetch(query, context);
            }
          };

      Future<Object> future = executor.submit(task);
        Object result = future.get(2, TimeUnit.MINUTES);
        data = (JSONObject) result;
```
where  `cF.cypherFetch(query, context);` is a function call where I connect to gremlin-server remotely and submit my query

**Comment by dwitry** [comment]
@SarthakGhosh16 solution you've posted will address the issue on the client side, but will not interrupt job on server, which may cause lags.

The best I can do to address your issue is to merge #316 into [tinkerpop334](https://github.com/opencypher/cypher-for-gremlin/tree/tinkerpop334) branch. This branch is not officially supported/released but it contains all changes 0.9.13 (before TinkerPop 3.4.x migration) + selected fixes. You can get it from [JitPack](https://jitpack.io/#opencypher/cypher-for-gremlin/tinkerpop334-SNAPSHOT).

The proper solution, of course, would be to upgrade JanusGraph 0.4.0/CfoG 1.0.4.

**Comment by SarthakGhosh16** [comment]
@dwitry Thanks for all your help. Since, we need official releases for our product, I don't think we can use releases from JitPack. 
I'll upgrade our client and server and use latest openCypher versions. 
Closing this issue.
